jobbergate-api/jobbergate_api/storage.py

```diff
@@ -1,24 +1,51 @@
 """
 Persistent data storage
 """
-import contextlib
 import re
 import typing
 
 import asyncpg
 import databases
+import databases.core
 import fastapi
 import pydantic
 import sqlalchemy
+from asyncpg.exceptions import UniqueViolationError
 from fastapi.exceptions import HTTPException
+from loguru import logger
 from sqlalchemy import Column, or_
 from sqlalchemy.sql.expression import BooleanClauseList, UnaryExpression
 from starlette import status
+from yarl import URL
 
-from jobbergate_api.config import settings
-from jobbergate_api.metadata import metadata
+from jobbergate_api.config import DeployEnvEnum, settings
 
-database = databases.Database(settings.DATABASE_URL)  # type: ignore
+INTEGRITY_CHECK_EXCEPTIONS = (UniqueViolationError,)
+
+
+def build_db_url(force_test: bool = False) -> str:
+    """
+    Helper method to build a database url based on settings. If ``is_test`` is passed, it will build from
+    the test database settings.
+
+    The ``force_test`` param allows you to override the check for the DEPLOY_ENV setting.
+    """
+    is_test = force_test or settings.DEPLOY_ENV == DeployEnvEnum.TEST
+    prefix = "TEST_" if is_test else ""
+
+    return str(
+        URL.build(
+            scheme="postgresql",
+            user=getattr(settings, f"{prefix}DATABASE_USER"),
+            password=getattr(settings, f"{prefix}DATABASE_PSWD"),
+            host=getattr(settings, f"{prefix}DATABASE_HOST"),
+            port=getattr(settings, f"{prefix}DATABASE_PORT"),
+            path="/{}".format(getattr(settings, f"{prefix}DATABASE_NAME")),
+        )
+    )
```

Comment:
I can't see the benefit of this module against having simply one variable in the config module defining the database URL. IMO we should just move the test postgres vars to the `docker-compose.yaml`

Reply:
This module also validates the URL.

Having separate settings for the test database makes it much safer. You have no chance of accidently connecting to a non-test database when running unit tests. If you don't separate them, there's a significant chance that you will accidently forget to set the database_url to connect to a test database.

Reply:
a test like [this one](https://github.com/omnivector-solutions/armada-api/blob/main/tests/conftest.py#L26-L33) could solve this risk.

Reply:
and that's where it converges with my previous comments: there are a bunch of variables to be set out there. A test like this one could make it simpler and then it was just needed to assign some dummy values as default for the database in the docker-compose.yaml file

Reply:
I don't think we should hard-code the expected test URL.

The biggest reason to break up the database settings is so that the same environment variables can be used to get values when creating the databases (via compose, the helper scripts, docker, whatever) as are used to connect within the app. This makes it much easier to just set your environment, start up your database, and know that your API is going to connect to that database.